### ISSUE TITLE
Validation of ContainerPorts

### ISSUE THREAD

**Issue by sayaoailun** [kubernetes/kubernetes#109498]
https://github.com/kubernetes/kubernetes/blob/c285e781331a3785a7f436042c65c5641ce8a9e9/pkg/apis/core/validation/validation.go#L2286-L2322

According above code, same port name in different container in the same pod is allowed, right?

e.g. `http` as the port name for container A and B both in the same pod.

https://github.com/kubernetes/kubernetes/blob/c285e781331a3785a7f436042c65c5641ce8a9e9/pkg/api/v1/pod/util.go#L30-L51

According above code, the first matched port will be chosen, right?

Should we do something to make it clear to avoid the use case?

**Comment by JayKayy** [comment]
I would be interested on working on this, after we get some direction on how to proceed.

**Comment by JayKayy** [comment]
@sayaoailun did you intend on making a PR for this or can I assign this to myself?

**Comment by sayaoailun** [comment]
@JayKayy I'm not making a PR. You are free to assign this to yourself.

**Comment by aojea** [comment]
interesting, the API docs says it should be unique per pod :thinking: 

https://github.com/kubernetes/kubernetes/blob/a750d8054a6cb3167f495829ce3e77ab0ccca48e/pkg/apis/core/types.go#L1800-L1805

in addition, `FindPort()` seems to be used by the endpoint/endpointslices controllers

```
pkg/controller/endpoint/endpoints_controller.go:                                portNum, err := podutil.FindPort(pod, servicePort)
pkg/controller/endpointslice/utils.go:          portNum, err := podutil.FindPort(pod, servicePort)
```

and the behavior is as reported

```diff
diff --git a/pkg/api/v1/pod/util_test.go b/pkg/api/v1/pod/util_test.go
index eb0425cf44a..a8ee7f8a244 100644
--- a/pkg/api/v1/pod/util_test.go
+++ b/pkg/api/v1/pod/util_test.go
@@ -121,6 +121,24 @@ func TestFindPort(t *testing.T) {
                port:     intstr.FromString("q"),
                expected: 22,
                pass:     true,
+       }, {
+               name: "valid str, two ctr with same name port",
+               containers: []v1.Container{{}, {Ports: []v1.ContainerPort{{
+                       Name:          "",
+                       ContainerPort: 11,
+                       Protocol:      "UDP",
+               }, {
+                       Name:          "q",
+                       ContainerPort: 22,
+                       Protocol:      "TCP",
+               }, {
+                       Name:          "q",
+                       ContainerPort: 33,
+                       Protocol:      "TCP",
+               }}}},
+               port:     intstr.FromString("q"),
+               expected: 22,
+               pass:     true,
        }, {
                name: "valid str, invalid protocol",
                containers: []v1.Container{{}, {Ports: []v1.ContainerPort{{
```

I don't think we can tight validation now, so I think that we should:
- fix the documentation
- add tests to assert the behavior

@thockin WDYT

**Comment by fedebongio** [comment]
/remove-sig api-machinery
We believe POD is owned by sig apps, therefore removing API Machinery. Thank you all!

**Comment by jlsong01** [comment]
@aojea @thockin Hi, if no object, I would like to fix the documentation and add tests according to @aojea 's comments  :-)

/assign

**Comment by thockin** [comment]
Yeah, this seems like a mistake.  We should comment it here as unfortunate, but required for compat.  Then we should add an API warning so people will at least see when they have done this.   That warning should appear any time someone does this in a Pod directly, but also in a Deployment, ReplicaSet, ReplicationController, Job, StatefulSet, etc.  I think we get that for free as validation calls into pod-sepc validation, but we should manually verify that.
